File: tracks/data_structures/queues/queries_with_fixed_length.cpp
#include <iostream>
#include <deque>
#include <vector>
#include <algorithm>
using namespace std;

template <typename T>
class MaxQueue {
  private:
    deque<T> items;
    deque<T> max;

  public:
    void append(T x) {
      items.push_back(x);

      while (!max.empty() && max.back() < x)
        max.pop_back();

      max.push_back(x);
    }

    T pop() {
      T x = items.front();
      items.pop_front();

      if (x == max.front())
        max.pop_front();

      return x;
    }

    T get_max() {
      return max.front();
    }
};


int main() {
  int n, q;
  vector<int> a;

  cin >> n >> q;

  int x;
  for (int i = 0; i < n; ++i) {
    cin >> x;
    a.push_back(x);
  }

  int d;
  for (int i = 0; i < q; ++i) {
    cin >> d;
    MaxQueue<int> queue;

    for (int j = 0; j < d; ++j) {
      queue.append(a[j]);
    }

    int m = queue.get_max();
    for (int j = d; j < n; ++j) {
      queue.pop();
      queue.append(a[j]);
      m = min(m, queue.get_max());
    }

    cout << m << endl;
  }

  return 0;
}

File: hackerrank/tracks/data_structures/queues/queries_with_fixed_length.py

from collections import deque


class MaxQueue(object):

    def __init__(self, a=None):
        self._q = deque()
        self._max = deque()

        if a:
            for x in a:
                self.append(x)


    def pop(self):
        x = self._q.pop()

        if self._max[-1] == x:
            self._max.pop()

        return x


    def append(self, x):
        self._q.appendleft(x)

        while self._max and x >= self._max[0]:
            self._max.popleft()
        self._max.appendleft(x)


    @property
    def max(self):
        return self._max[-1]


n, q = [int(x) for x in raw_input().split()]
a = [int(x) for x in raw_input().split()]

for _ in range(q):
    d = int(raw_input())

    queue = MaxQueue(a[:d])
    res = queue.max

    for x in a[d:]:
        queue.pop()
        queue.append(x)
        res = min(res, queue.max)

    print res
